Pl 127635 stabilize network reconfiguration

ctheune opened the pull request:
@flyingcircusio/release-managers

## Release process

Impact:

* All network interfaces will be restarted, causing a short service interruption of a few seconds. (PL-127635)

Changelog:

* We have rebuilt a major part of the network configuration logic to allow more fine grained online network reconfiguration in the future and to make it a lot more robust when making changes like changing IP addresses, renaming interfaces  and shifting between more complex architectures like tagged networks, bridge setups and even EVPN-VXLAN environments.

### PR release workflow (internal)

- [x] PR has internal ticket
- [x] internal issue ID (PL-…) part of branch name
- [x] internal issue ID mentioned in PR description text
- [x] ticket is on Platform agile board
- [ ] ticket state set to *Pull request ready*
- [ ] if ticket is more urgent than within the next few days, directly contact a member of the Platform team

## Design notes

- [x] Provide a feature toggle if the change might need to be adjusted/reverted quickly depending on context. Consider whether the default should be `on` or `off`. Example: rate limiting.
- [x] All customer-facing features and (NixOS) options need to be discoverable from documentation. Add or update relevant documentation such that hosted and guided customers can understand it as well.

## Security implications

- [x] [Security requirements](https://wiki.flyingcircus.io/System_Development_Guideline#Security_requirement_principles_and_testing) defined? (WHERE)

Ensure availability by improving robustness.

- [x] Security requirements tested? (EVIDENCE)

Relying on the existing automated tests. Also extensively tested on cartman06 (Ceph) while developing and tested on cartman37 and kyle05 and tstest21 to check production rollout impact. In online scenarios this causes about 10-15 seconds of interruption in EVPN-VXLAN situations, less than 5 seconds in non-EVPN-VXLAN setups and VMs.

ctheune committed:
logging: optimize serial console noise on hardware, increase log size

ctheune committed:
lib/network: fix relaxedIP script to assist json parsing

ctheune committed:
network: don't make lldp altname fail completely on individual errors

ctheune committed:
network: revamp interface to make reconfiguration more robust

1. Do not tear down the whole "house of cards" as @mm likes to call it.

2. Move away from relying on systemd device units and udev for network
   interfaces and their naming, because they are extremely error prone
   under renaming situations.

   Provide a more robust (non-udev) approach to renaming network
   interfaces including managing altnames.

   This folds a number of little snippets into a more generic
   `-netdev.service` units.

   Do not use partOf but rely on a combination of
   wantedBy=network-setup.service and requires=network-setup.service to
   make the house of cards more stable.

3. Include a (development) version of our nixos-21.05 fork that
   homogenizes points 1 and 2 with the nixos foundational parts.

3

sysvinit commented on nixos/platform/network.nix:
```diff
@@ -357,63 +350,115 @@ in
+                echo "Got interface renaming lock ..."
+
+                # Look up the desired interface's current name by mac address
+                interface="$(ip -j -d link show | jq -r '.[] | select(.linkinfo?.info_kind? == null) | select(.address == "${iface.mac}") | .ifname')"
```
Another thought which comes to mind: what happens if there is no interface with the desired MAC address at all? In the world of `.device` units, this would cause dependents to fail with a dependency error (after a timeout?) - presumably this script would just exit with an error somewhere.

(Handling this better could reasonably be kicked down the road to the discussed Python implementation...)

ctheune replied:
will do, made a note in the script so i won't forget during python conversion.

sysvinit commented on nixos/platform/network.nix:
```diff
@@ -532,13 +574,31 @@ in
+          # NixOS scripted networking configuration does not know
+          # about VXLAN devices, and (incorrectly) generates
+          # dependencies for them as if they were physical ethernet
+          # devices. We need to patch some dependencies in other units
+          # so that the link configuration and bridge device creation
+          # depend on the correct units.
+
           (map (iface: (lib.nameValuePair
             "network-addresses-${iface.interface}"
-            rec {
-              after = [ "${iface.link}-netdev.service" ];
-              bindsTo = after;
+            {
+              after = [ "network-pre.target" "${iface.link}-netdev.service" "zebra.service" ];
+              bindsTo = [ "${iface.link}-netdev.service" ];
             }
           )) vxlanInterfaces) ++
+
+          (map (iface: (lib.nameValuePair
+            "${iface.interface}-netdev"
+            {
+              after = [ "network-pre.target" "network-addresses-${iface.link}.service" ];
+              # I'd love to bind those to zebra, but that creates a cycle.
+              bindsTo = [ "network-addresses-${iface.link}.service" ];
+            }
+          )) vxlanInterfaces) ++
+
```
I originally wrote these overrides because NixOS was incorrectly generating dependencies on `.device` units for virtual interfaces where this makes no sense.

As this branch is using a patched version of nixpkgs which gets rid of device unit dependencies completely, these overrides specify redundant dependencies (e.g. the generated unit file for `brfe-netdev.service` has a duplicated After= dependency on `network-addresses-vxfe.service`).

Both of those overrides can be removed, if we're not making any further local customisations to dependencies for these interfaces.

sysvinit replied:
Same applies to the override at line 521 for `network-addresses-ul-loopback.service`, this can also be removed.

ctheune replied:
yup, those are right. reviewed together.

ctheune committed:
network: more cleanups and test fixes for the new network unit code

ctheune committed:
network: make the new externalLabel optional again, fix resolver tests

I'd love for the externalLabel to not be optional, but our test
suite isn't flexible enough to adapt to that and I'd have to manually
inspect and change roughly a bazillion tests ... so lets roll with a
default for now.

ctheune committed:
network: fix sysctls after logging error review

ctheune committed:
network: new nixpkgs for backport to udev issue

ctheune committed:
typo

ctheune committed:
network: more typos

ctheune committed:
another nixpkgs update to fix type issue in deviceDependencies
